### PR TITLE
feat(server): random assets API

### PR DESCRIPTION
**Pull request by daniele-athome**
Adding a new API for assets:

```
/api/asset/random?count=5
```

"count" parameter is optional, defaulting to 1.

It will return `count` random assets owned by the authenticated user.

Possible things for the review:

* automated tests: I didn't know what to test here, no special logic in the service, just doing a SQL query
* selection query: I had to use a custom SQL query - see comments inside
* I did some research and tests and I picked the fastest and most reliable methods among the ones I found

## How was this tested?

* called API with no parameters -> defaults to 1 asset
* called API with invalid parameter -> HTTP 400
* called API with valid parameter -> returns N assets

### CODE REVIEW

File: server/src/domain/asset/dto/asset.dto.ts [hunk 1]
```diff
@@ -25,3 +26,11 @@ export class UpdateAssetDto {
   @IsString()
   description?: string;
 }
+
+export class RandomAssetsDto {
+  @Optional()
+  @IsNumber()
```
Comment (jrasm91) [inline server/src/domain/asset/dto/asset.dto.ts:32]: Is int

File: server/src/domain/asset/dto/asset.dto.ts [hunk 2]
```diff
@@ -25,3 +26,11 @@ export class UpdateAssetDto {
...
+  @IsPositive()
+  @Type(() => Number)
+  count?: number | null;
```
Comment (jrasm91) [inline server/src/domain/asset/dto/asset.dto.ts:35]: Null is not valid here, so you can remove that type.

File: server/src/infra/repositories/asset.repository.ts [hunk 3]
```diff
@@ -429,6 +429,17 @@ export class AssetRepository implements IAssetRepository {
     return result;
   }
 
+  getRandom(ownerId: string, count: number): Promise<AssetEntity[]> {
+    // can't use queryBuilder because of custom OFFSET clause
+    return this.repository.query(
+      `SELECT *
+                                  FROM assets
+                                  WHERE "ownerId" = $1
+                                  OFFSET FLOOR(RANDOM() * (SELECT COUNT(*) FROM ASSETS)) LIMIT $2`,
```
Comment (jrasm91) [inline server/src/infra/repositories/asset.repository.ts:438]: Looks like this could return less than the limit in a few situations if the offset was close to the number of assets. Probably not a big deal for most situations though.

**Comment by jrasm91** [comment]
Can you add a few e2e tests with the situations you described? Should be easy enough to add.

File: server/src/infra/repositories/asset.repository.ts (hunk 3 above)
Comment (daniele-athome) [reply server/src/infra/repositories/asset.repository.ts]: You're right, I didn't think about boundary values. I think I fixed it in a more or less proper way.

**Comment by daniele-athome** [comment]
I implemented tests for all use cases, although I didn't find a way to properly `expect` a random asset in a "jesty" way - there seems to be extensions for jest that could implement those, but I didn't think it was worth to add a dependecy just for an OR check.

I think I also found a typo in another test - I isolated the fix in a separate commit.

File: server/test/e2e/asset.e2e-spec.ts [hunk 4]
```diff
@@ -378,6 +385,58 @@ describe(`${AssetController.name} (e2e)`, () => {
...
+        .get('/asset/random')
+        .set('Authorization', `Bearer ${user1.accessToken}`);
+
+      expect(status).toBe(200);
+
+      const assets: AssetResponseDto[] = body;
+      expect(assets.length).toBe(1);
+      expect(assets[0].ownerId).toBe(user1.userId);
+      // assets owned by user1
+      expect(assets[0].id == asset1.id || assets[0].id == asset2.id || assets[0].id == asset3.id).toBeTruthy();
```
Comment (danieldietzler) [inline server/test/e2e/asset.e2e-spec.ts:407]: I personally would prefer `expect([asset1.id, asset2.id, asset3.id]).toContain(assets[0].id)`

File: server/test/e2e/asset.e2e-spec.ts [hunk 5]
```diff
@@ -378,6 +385,58 @@ describe(`${AssetController.name} (e2e)`, () => {
...
+      // assets owned by user2
+      expect(assets[0].id == asset4.id).toBeFalsy();
```
Comment (danieldietzler) [inline server/test/e2e/asset.e2e-spec.ts:409]: You can do `expect(asset[0].id).not.toBe(asset4.id)` instead

File: server/test/e2e/asset.e2e-spec.ts [hunk 6]
```diff
@@ -322,7 +329,7 @@ describe(`${AssetController.name} (e2e)`, () => {
     });
 
     it('should require authentication', async () => {
-      const { status, body } = await request(server).get('/album/statistics');
+      const { status, body } = await request(server).get('/asset/statistics');
```
Comment (danieldietzler) [inline server/test/e2e/asset.e2e-spec.ts:332]: Thanks for the good catch! :) That was my bad.

File: server/test/e2e/asset.e2e-spec.ts (hunk 4 above)
Comment (daniele-athome) [reply server/test/e2e/asset.e2e-spec.ts:407]: I don't know why I didn't think of that, thank you :-)

Comment (danieldietzler) [reply server/test/e2e/asset.e2e-spec.ts]: Well, it is somewhat counterintuitive :D

**Review by danieldietzler — APPROVED** [review]
LGTM! I would like to have Jason take another look at it though.

**Review by alextran1502 — APPROVED** [review]
